test/common/stats/thread_local_store_test.cc

```diff
@@ -343,5 +453,135 @@ TEST_F(StatsThreadLocalStoreTest, ShuttingDown) {
   EXPECT_CALL(*this, free(_)).Times(5);
 }
 
+// Histogram tests
+TEST_F(HistogramTest, BasicSingleHistogramMerge) {
+  Histogram& h1 = store_->histogram("h1");
+  EXPECT_EQ("h1", h1.name());
+
+  expectCallAndAccumulate(h1, 0);
+  expectCallAndAccumulate(h1, 43);
+  expectCallAndAccumulate(h1, 41);
+  expectCallAndAccumulate(h1, 415);
+  expectCallAndAccumulate(h1, 2201);
+  expectCallAndAccumulate(h1, 3201);
+  expectCallAndAccumulate(h1, 125);
+  expectCallAndAccumulate(h1, 13);
+
+  EXPECT_EQ(1, validateMerge());
+}
+
+TEST_F(HistogramTest, BasicMultiHistogramMerge) {
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = store_->histogram("h2");
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("h2", h2.name());
+
+  expectCallAndAccumulate(h1, 1);
+  expectCallAndAccumulate(h2, 1);
+  expectCallAndAccumulate(h2, 2);
+
+  EXPECT_EQ(2, validateMerge());
+}
+
+TEST_F(HistogramTest, MultiHistogramMultipleMerges) {
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = store_->histogram("h2");
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("h2", h2.name());
+
+  // Insert one value in to one histogram and validate
+  expectCallAndAccumulate(h1, 1);
+  EXPECT_EQ(2, validateMerge());
+
+  // Insert value into second histogram and validate that it is merged properly.
+  expectCallAndAccumulate(h2, 1);
+  EXPECT_EQ(2, validateMerge());
+
+  // Insert more values into both the histograms and validate that it is merged properly.
+  expectCallAndAccumulate(h1, 2);
+  EXPECT_EQ(2, validateMerge());
+
+  expectCallAndAccumulate(h2, 3);
+  EXPECT_EQ(2, validateMerge());
+
+  expectCallAndAccumulate(h2, 2);
+  EXPECT_EQ(2, validateMerge());
+
+  // Do not insert any value and validate that intervalSummary is empty for both the histograms and
+  // cumulativeSummary has right values.
+  EXPECT_EQ(2, validateMerge());
+}
+
+TEST_F(HistogramTest, BasicScopeHistogramMerge) {
+  ScopePtr scope1 = store_->createScope("scope1.");
+
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = scope1->histogram("h2");
+  EXPECT_EQ("h1", h1.name());
+  EXPECT_EQ("scope1.h2", h2.name());
+
+  expectCallAndAccumulate(h1, 2);
+  expectCallAndAccumulate(h2, 2);
+  EXPECT_EQ(2, validateMerge());
+}
+
+TEST_F(HistogramTest, BasicHistogramValidate) {
+  Histogram& h1 = store_->histogram("h1");
+  Histogram& h2 = store_->histogram("h2");
+
+  expectCallAndAccumulate(h1, 1);
+
+  EXPECT_EQ(2, validateMerge());
+
+  std::string h1_expected_summary =
+      "P0: 1, P25: 1.025, P50: 1.05, P75: 1.075, P90: 1.09, P95: 1.095, "
+      "P99: 1.099, P99.9: 1.0999, P100: 1.1";
+  std::string h2_expected_summary =
```

Comment:
make these locals const. or just move the literals into their single use. 

BTW I'm not really sure why you need the degenerate case; seems like you are testing behavior which is really at the boundary of sanity, and probably not that important.  But it's nice to see that the even-distribution case is working.

Reply:
i want to verify that case as well. I will leave it like that.